### PR TITLE
In uniqnum: Avoid reading of unused bytes

### PR DESCRIPTION
**Pull request by sisyphus**
First up, a small fix to something I should have noticed earlier:
Change line 1395 of current github version of ListUtil.xs from:

sv_setpvn(keysv, (char *) &nv_arg, sizeof(NV));
to
sv_setpvn(keysv, (char *) &nv_arg, 8);

This is because, at that point, sizeof(NV) is always 8.

The remaining changes address my concern about reading in all 12 (or 16, as the case might be) of the 80-bit extended precision long double when only 10 are needed.
Firstly, why go to the trouble of reading 12/16 bytes when only 10 are relevant.
Secondly, I have concerns that the bytes beyond the 10th might contain random garbage that would lead to 2 equivalent long doubles being deemed to have unequal values.

### CODE REVIEW

**Comment by sisyphus** [comment]
BTW, I don't have any additional PRs to make in the immediate future.

It'd be great if we could get this PR applied and a new Scalar-List-Utils release made so that we can subject it to wider testing and fix any issues in time for perl-5.32.0.

Cheers,
Rob

**Comment by sisyphus** [comment]
I've come up with a much cleaner (and slightly faster) way of dealing with the case where IVSIZE == NVSIZE == 8.
(I hadn't expected to come up with such an approach so soon, but it turned out to be easier than I had anticipated. So I'll outline it here, but I won't submit a PR for it until after this one has been dealt with.)
Now would be a good time for comments upon it, and especially for any ideas that would improve it.

Basically, it enables us to do away with the Windows-specific stuff  - there's no longer a need to detect mingw compilers, nor do we need to define __USE_MINGW_ANSI_STDIO, and it also works with *all* microsoft compilers - not just the recent ones.
Anyway ... if there's anyone interested, take a look at the attached new_stuff.txt

[new_stuff.txt](https://github.com/Dual-Life/Scalar-List-Utils/files/4249864/new_stuff.txt)

**Comment by sisyphus** [comment]
Tonight, David (who else) Mitchell presented a simpler method of determining whether a UV can be represented exactly as an NV at:
https://www.perlmonks.org/?node_id=11113426

I'll go with his suggestion when I make that next PR.

Cheers,
Rob

**Comment by sisyphus** [comment]
It eventually turned out that Dave Mitchell's idea for a solution is susceptible to the very problem I was trying to work around.
Anyway ... enough with speculation about what I want to change next ... could we just get this PR applied ... or get some feedback on what's unacceptable about it.

I'm starting to worry about the appalling possibility of perl-5.32.0 core shipping with either S-L-U-1.53 or S-L-U 1.54.

Cheers,
Rob

**Comment by leonerd** [comment]
Seems a bit fragile, but probably no more fragile than it was originally.

**Comment by sisyphus** [comment]
In terms of how it's testing (List-Uniqnum-0.05) it seems quite robust:
453 ALLs 452 PASSes 1 UNKNOWN.

The one "UNKNOWN" makes no sense to me. I doubt very much that it has anything to do with the actual XS code.

The only possible point of fragility I can see is its reliance on the C toolchain to correctly cast an IV/UV to an NV.
And, remember, such casts are being performed *only* upon IV/UVs that can be *exactly* represented as an NV.
IIRC, all other IV/UVs are being stringified with this particular code.
If the C toolchain can't correctly cast such an IV/UV to an NV then I think there are bigger concerns than just Lit::Util::uniqnum.

Anyway, I'd better investigate the reason that perl-5.6.2 is choking on #100
